### DISCUSSION TITLE
TypeScript icon missing on packages which bundle declarations and do not need a "types" field

### DISCUSSION THREAD

**Discussion by cspotcode** [npm/feedback#122] [Website] [answered]
Related to #64

npm's new TS icon is not appearing for packages which bundle their own declarations in a way that does not need a "types" field.

Here's an example of a package which does this and is not getting the icon.
https://www.npmjs.com/package/execa
https://unpkg.com/execa@5.0.0/index.d.ts

The types are located at `index.d.ts`, so a `types` field is not necessary.

This package is not covered by the 3x categories mentioned in npm/roadmap#20, but the first category should be broadened to include it.

Another way this can happen, if `"main"` points both to JS and DTS.

```
  "main": "./dist"
```

The compiler by default emits `.js` and `.d.ts` files side-by-side, so this package.json satisfies both node and TypeScript resolvers, pointing to `./dist/index.js` and `./dist/index.d.ts`

Here's an example of this:
https://www.npmjs.com/package/clipanion
https://unpkg.com/browse/clipanion@2.6.2/package.json
https://unpkg.com/browse/clipanion@2.6.2/lib/index.d.ts

TypeScript's documentation for this behavior:
https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html#editing-the-packagejson

**Comment by orta** [comment]
Linking to https://github.com/npm/rfcs/pull/126 which would add the "types" field in a package (though not retrospectively)  meaning less future work

**Comment by MylesBorins** [comment]
As per https://twitter.com/ethomson/status/1339279153116360714 supporting '.d.ts' is on the radar.

**Reply by panva** [reply]
@MylesBorins is any `.d.ts` file presence going to flip the icon on? The tweet seems to suggest that the files would have to be next to the `.js` files, which isn't always the case.

Case in point http://npmjs.com/package/jose, a package bundling multiple ESM modules using Conditional Subpath Exports. In order for TS to be able to resolve the types (since TS itself does not support Subpath Exports at all) and not having to duplicate them for each distribution, i use the `typesVersions` field to point the TS resolver in the right place regardless of which module or its conditional variant gets required/imported.

**Reply by ethomson** [reply]
Has TypeScript documented `typesVersions`?  That would be the first place to start.  We're implementing their spec, not breaking new ground.

**Reply by panva** [reply]
Yep, https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#version-selection-with-typesversions

> When TypeScript opens a package.json file to figure out which files it needs to read, it first looks at a new field called typesVersions.

<details>
  <summary><em><strong>Example resolution:</strong></em> (Click to expand)</summary>

```
======== Resolving module 'jose/jwe/compact' from '/Users/panva/repo/esm/some.ts'. ========
Explicitly specified module resolution kind: 'NodeJs'.
Loading module 'jose/jwe/compact' from 'node_modules' folder, target file type 'TypeScript'.
Found 'package.json' at '/Users/panva/repo/esm/node_modules/jose/package.json'.
'package.json' has a 'typesVersions' field with version-specific path mappings.
'package.json' has a 'typesVersions' entry '*' that matches compiler version '4.0.3', looking for a pattern to match module name 'jwe/compact'.
Module name 'jwe/compact', matched pattern '*'.
Trying substitution './types/*', candidate module location: './types/jwe/compact'.
File '/Users/panva/repo/esm/node_modules/jose/types/jwe/compact.d.ts' exist - use it as a name resolution result.
Resolving real path for '/Users/panva/repo/esm/node_modules/jose/types/jwe/compact.d.ts', result '/Users/panva/repo/esm/node_modules/jose/types/jwe/compact.d.ts'.
======== Module name 'jose/jwe/compact' was successfully resolved to '/Users/panva/repo/esm/node_modules/jose/types/jwe/compact.d.ts' with Package ID 'jose/types/jwe/compact.d.ts@1.0.0'. ========
```
</details>

As you can see at no point did TypeScript check the `types` field, the use of `typesVersions` and a direct file name match in its resolved path comes first.

**Reply by ethomson** [reply]
Thanks for the link.  Indeed we'll add that support.

> As you can see at no point did TypeScript check the types field

That's documented in the first paragraph of the same document: https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#including-declarations-in-your-npm-package

**Reply by panva** [reply]
@ethomson is there a new roadmap item / discussion to track the progress on `typesVersions` and other means of identifying that types are included?

**Comment by ethomson** [comment, marked as answer]
Hi @cspotcode and @panva - we should now identify both `.d.ts` files that map to a specific `main` and types as specified by `typesVersions`.  Thanks for your feedback!

**Reply by panva** [reply]
I can confirm. Thank you!

**Comment by DetachHead** [comment]
something seems to have changed recently because my package no longer has the icon: https://www.npmjs.com/package/@detachhead/ts-helpers

but it used to have the icon even though it doesn't have/need a `"types"` or `"main"` field

**Reply by DetachHead** [reply]
i did this to work around it https://github.com/DetachHead/ts-helpers/commit/72bf00d6a7d21c2554be1c5ba56f72593aad3c55
